File: scripts/get-wcb-tasks.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
AI × Web3 School - WCB 任务获取脚本
使用 API 密钥从 web3career.build 获取课程任务
"""

import requests
import json
import os
from pathlib import Path

class WCBTaskFetcher:
    """WCB 任务获取器"""

    def __init__(self, api_key):
        self.api_key = api_key
        self.base_url = "https://web3career.build"
        self.headers = {
            "Authorization": f"Bearer {api_key}",
            "Content-Type": "application/json"
        }
        self.tasks = []

    def get_tasks(self):
        """获取 AI × Web3 School 的所有任务"""
        try:
            # API 端点（基于之前的搜索结果）
            url = f"{self.base_url}/api/v1/tasks"

            # 发送 GET 请求
            response = requests.get(url, headers=self.headers)

            if response.status_code == 200:
                data = response.json()
                self.tasks = data.get('tasks', [])
                print(f"✅ 成功获取 {len(self.tasks)} 个任务")
                return self.tasks
            else:
                print(f"❌ 获取任务失败: {response.status_code}")
                print(f"错误信息: {response.text}")
                return []

        except requests.exceptions.RequestException as e:
            print(f"❌ 网络错误: {e}")
            return []

    def extract_task_names(self):
        """提取任务名称"""
        task_names = []
        for task in self.task:
            # 根据任务结构提取名称
            name = task.get('title', task.get('name', 'Unknown Task'))
            task_names.append(name)
        return task_names

    def create_task_folders(self, task_names, base_path):
        """根据任务名称创建文件夹"""
        created_folders = []

        for task_name in task_names:
            # 清理任务名称，使其成为有效的文件夹名
            folder_name = self.clean_folder_name(task_name)
            folder_path = Path(base_path) / folder_name

            # 检查文件夹是否已存在
            if not folder_path.exists():
                folder_path.mkdir(parents=True, exist_ok=True)
                print(f"📁 创建文件夹: {folder_path}")
                created_folders.append({
                    'name': task_name,
                    'path': str(folder_path)
                })
            else:
                print(f"📁 文件夹已存在: {folder_path}")

        return created_folders

    def clean_folder_name(self, name):
        """清理文件名，移除特殊字符"""
        # 替换特殊字符为下划线
        import re
        name = re.sub(r'[^\w\s-]', '_', name)
        # 替换多个下划线为单个
        name = re.sub(r'[_]+', '_', name)
        # 去除首尾下划线
        name = name.strip('_')
        # 限制长度
        if len(name) > 50:
            name = name[:50]
        return name

    def create_task_readme(self, task_info, folder_path):
        """为每个任务创建 README 文件"""
        readme_content = f"""# {task_info.get('name', 'Task')}

## 任务描述
{task_info.get('description', '暂无描述')}

## 任务要求
{self.format_requirements(task_info.get('requirements', []))}

## 截止日期
{task_info.get('due_date', '未设定')}

## 相关资源
{self.format_resources(task_info.get('resources', []))}

## 进度跟踪
- [ ] 开始任务
- [ ] 进行中
- [ ] 提交完成
- [ ] 获得反馈

---

*创建时间: {task_info.get('created_at', '未知')}*
*任务ID: {task_info.get('id', '未知')}*
"""

        readme_path = Path(folder_path) / "README.md"
        with open(readme_path, 'w', encoding='utf-8') as f:
            f.write(readme_content)
        print(f"📝 创建 README: {readme_path}")

def main():
    """主函数"""
    # API 密钥
    api_key = "w3cb_sk_m-TCyb71gQeg-xh1La_L_Rq0ViFvkh7m"

    # 初始化任务获取器
    fetcher = WCBTaskFetcher(api_key)

    # 获取任务
    tasks = fetcher.get_tasks()

    if tasks:
        # 提取任务名称
        task_names = fetcher.extract_task_names()

        # 设置基础路径
        base_path = Path("/e/Ethereum/AI x Web3 School/ai-web3-school-cohort-rain/tasks")

        # 创建任务文件夹
        created_folders = fetcher.create_task_folders(task_names, base_path)

        print(f"\n📊 总结:")
        print(f"- 总任务数: {len(tasks)}")
        print(f"- 新建文件夹: {len(created_folders)}")
        print(f"- 已存在文件夹: {len(task_names) - len(created_folders)}")

        # 为每个创建的文件夹创建 README
        for folder_info in created_folders:
            # 找到对应的任务详情
            task_detail = next((t for t in tasks if t.get('title') == folder_info['name']), None)
            if task_detail:
                fetcher.create_task_readme(task_detail, folder_info['path'])
    else:
        print("❌ 未能获取到任务信息")

if __name__ == "__main__":
    main()

File: scripts/get-wcb-tasks.js
/**
 * AI × Web3 School - WCB 任务获取脚本
 * 使用 API 密钥从 web3career.build 获取课程任务
 */

const fs = require('fs');
const path = require('path');

class WCBTaskFetcher {
    constructor(apiKey) {
        this.apiKey = apiKey;
        this.baseUrl = 'https://web3career.build';
        this.headers = {
            'Authorization': `Bearer ${apiKey}`,
            'Content-Type': 'application/json'
        };
        this.tasks = [];
    }

    async getTasks() {
        try {
            // API 端点
            const url = `${this.baseUrl}/api/v1/tasks`;

            // 发送请求
            const response = await fetch(url, {
                method: 'GET',
                headers: this.headers
            });

            if (response.ok) {
                const data = await response.json();
                this.tasks = data.tasks || [];
                console.log(`✅ 成功获取 ${this.tasks.length} 个任务`);
                return this.tasks;
            } else {
                console.error(`❌ 获取任务失败: ${response.status}`);
                const errorText = await response.text();
                console.error(`错误信息: ${errorText}`);
                return [];
            }
        } catch (error) {
            console.error(`❌ 网络错误: ${error.message}`);
            return [];
        }
    }

    extractTaskNames() {
        return this.tasks.map(task => {
            return task.title || task.name || 'Unknown Task';
        });
    }

    createTaskFolders(taskNames, basePath) {
        const createdFolders = [];
        const tasksDir = path.join(basePath, 'tasks');

        // 确保任务目录存在
        if (!fs.existsSync(tasksDir)) {
            fs.mkdirSync(tasksDir, { recursive: true });
        }

        taskNames.forEach(taskName => {
            const folderName = this.cleanFolderName(taskName);
            const folderPath = path.join(tasksDir, folderName);

            // 检查文件夹是否存在
            if (!fs.existsSync(folderPath)) {
                fs.mkdirSync(folderPath, { recursive: true });
                console.log(`📁 创建文件夹: ${folderPath}`);
                createdFolders.push({
                    name: taskName,
                    path: folderPath
                });
            } else {
                console.log(`📁 文件夹已存在: ${folderPath}`);
            }
        });

        return createdFolders;
    }

    cleanFolderName(name) {
        // 清理文件名，移除特殊字符
        let cleaned = name
            .replace(/[^\w\s-]/g, '_')  // 替换特殊字符
            .replace(/[_]+/g, '_')      // 合并多个下划线
            .replace(/^_|_$/g, '')      // 去除首尾下划线
            .substring(0, 50);          // 限制长度

        return cleaned;
    }

    createTaskReadme(taskInfo, folderPath) {
        const readmeContent = `# ${taskInfo.name || 'Task'}

## 任务描述
${taskInfo.description || '暂无描述'}

## 任务要求
${this.formatRequirements(taskInfo.requirements || [])}

## 截止日期
${taskInfo.due_date || '未设定'}

## 相关资源
${this.formatResources(taskInfo.resources || [])}

## 进度跟踪
- [ ] 开始任务
- [ ] 进行中
- [ ] 提交完成
- [ ] 获得反馈

---

*创建时间: ${taskInfo.created_at || '未知'}*
*任务ID: ${taskInfo.id || '未知'}*
`;

        const readmePath = path.join(folderPath, 'README.md');
        fs.writeFileSync(readmePath, readmeContent, 'utf8');
        console.log(`📝 创建 README: ${readmePath}`);
    }

    formatRequirements(requirements) {
        if (!requirements || requirements.length === 0) {
            return '暂无具体要求';
        }

        return requirements.map(req => `- ${req}`).join('\n');
    }

    formatResources(resources) {
        if (!resources || resources.length === 0) {
            return '暂无相关资源';
        }

        return resources.map(resource => `- [${resource.name || resource.title}](${resource.url || '#'})`).join('\n');
    }

    async run() {
        try {
            // 获取任务
            const tasks = await this.getTasks();

            if (tasks.length > 0) {
                // 提取任务名称
                const taskNames = this.extractTaskNames();

                // 设置基础路径
                const basePath = path.join(__dirname, '..');

                // 创建任务文件夹
                const createdFolders = this.createTaskFolders(taskNames, basePath);

                console.log('\n📊 总结:');
                console.log(`- 总任务数: ${tasks.length}`);
                console.log(`- 新建文件夹: ${createdFolders.length}`);
                console.log(`- 已存在文件夹: ${taskNames.length - createdFolders.length}`);

                // 为每个创建的文件夹创建 README
                for (const folderInfo of createdFolders) {
                    const taskDetail = tasks.find(t =>
                        (t.title || t.name) === folderInfo.name
                    );
                    if (taskDetail) {
                        this.createTaskReadme(taskDetail, folderInfo.path);
                    }
                }
            } else {
                console.log('❌ 未能获取到任务信息');
            }
        } catch (error) {
            console.error('❌ 执行出错:', error.message);
        }
    }
}

// 主函数
async function main() {
    const apiKey = 'w3cb_sk_m-TCyb71gQeg-xh1La_L_Rq0ViFvkh7m';

    const fetcher = new WCBTaskFetcher(apiKey);
    await fetcher.run();
}

// 运行主函数
if (require.main === module) {
    main();
}

module.exports = WCBTaskFetcher;